### ISSUE TITLE
InvalidRegistration iOS Push Notification

### ISSUE THREAD

**Issue by alekaizer** [expo/expo#13949]
### Summary

Push Notification generated from iOS is not working

### Environment

  Expo CLI 4.8.1 environment info:
    System:
      OS: macOS 10.15.2
      Shell: 5.7.1 - /bin/zsh
    Binaries:
      Node: 14.17.3 - /usr/local/bin/node
      Yarn: 1.3.2 - /opt/local/bin/yarn
      npm: 6.14.13 - /usr/local/bin/npm
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 13.2, DriverKit 19.0, macOS 10.15, tvOS 13.2, watchOS 6.1
      Android SDK:
        API Levels: 27, 29
        Build Tools: 29.0.2
    IDEs:
      Android Studio: 3.6 AI-192.7142.36.36.6392135
      Xcode: 11.3/11C29 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~40.0.0 => 40.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz => 0.63.2 
    npmGlobalPackages:
      expo-cli: 4.8.1
    Expo Workflow: managed

### Please specify your device/emulator/simulator platform, model and version

iPhone 7/iPhone 11 Pro

### Error output

{"multicast_id":6161497705980646930,"success":0,"failure":1,"canonical_ids":0,"results":[{"error":"InvalidRegistration"}]}

### Reproducible demo or steps to reproduce from a blank project

I’ve built and released the app on the Play Store following the FCM guide, meaning

- getting the json config for the app to get FCM token
- getting server key to be able to send notifications from the server
- all of it worked like a charm

Then when I got to iOS, followed the guide as well

- got the PLIST file
- added the file to the app and built it
- when configuring the app on Firebase console, I got asked the .p8 file, the one generated/added by Expo was no longer downloadable
- I added a new one and replaced the one on expo using the credentials manager
- replaced the provision profile as well
- Built the app, released it

Token are generated, up to this point everything seems fine
Now trying to send message to the iOS device always resulted in the same response

 `{"multicast_id":6161497705980646930,"success":0,"failure":1,"canonical_ids":0,"results":[{"error":"InvalidRegistration"}]}`

Here are the steps I followed

### On Apple Dev Account

- Added a key
- Added a Provision Profile
- Added both on Expo Credentials(verified all the IDs, everything checks out)
- On Firebase Console
- 

### Added iOS app

- Fill in the Apple informations(Team ID, bundle, etc…)
- Downloaded the plist file
- Uploaded the .p8 file
- Copied the Server key to the backend server

### On the App

   - added the plist file path in the app.json
   - built the app
   - fetched the device token(seems legit, 64 chars long)

But when I try to send a notification to the iOS devices it doesn’t work, but on Android everything works fine.

**Comment by cruzach** [comment]
hey @alekaizer , can you explain what exactly you're trying to do? In your steps for iOS you mention "when configuring the app on Firebase console"- iOS notifications from Expo are not done using FCM, that is only for Android. 

Are you sending notifications with Expo's notification service, or with your own? Although you've provided lots of details, what would be more helpful is a short list of steps to reproduce the same issue

**Comment by alekaizer** [comment]
@cruzach thanks for replying. The thing is I don't know if you will be able to reproduce it, but the steps I gave above are pretty much it

What I'm trying to achieve is that, since FCM support APNs, I can configure Android and iOS on the Firebase Console
- add an iOS app
- follow the wizard 
- download the plist(which is just like the json for Android)
- add the APNs Authentication Key to the app on the firebase console

Pretty much what is described [here(Set up FCM client app on iOS)](https://firebase.google.com/docs/cloud-messaging/ios/client)

So my expectation was that doing that, I could get the push token and from the server send a notification to the device using the FCM Key just like I did for Android, but it seems to not be the case

Before your message I was contemplating switching to Expo push token, but it doesn't seem to accept topics, so I can broadcast messages to a topic instead of fetching all the tokens one by one.


In summary, what I was expecting was to be able to use FCM ability to support iOS to send notification to both platforms, without using expo notification sdk or writing a custom flow/logic for APNs

I hope I was clear in my explanation

**Comment by cruzach** [comment]
Aha yeah I see what you're trying to do now, have you searched through FCMs documentation for that error message? If it's the response you're getting from FCM, then you should refer to their docs (you're just using expo-notifications to get the device push token)

**Comment by alekaizer** [comment]
Yes, I checked the error didn't find anything that could help, maybe it is because I was too focus on it being from iOS.
Lemme dig into a bit and get back to you, hopefully I find a solution with a clearer head

**Comment by alekaizer** [comment]
@cruzach I think I might have found the issue, when I call `getDevicePushTokenAsync()` on iOS, what I'm getting is a APNs token, right?
If that's the case I would need to convert it to FCM token for FCM to be able to send the notification. 

maybe something worth adding to `expo-notification` a method like `registerForRemoteNotification()`

**Comment by cruzach** [comment]
Correct, on iOS you're getting the native iOS token. Unless you are comfortable with APNS and FCM, I would recommend you try out Expo's push notification service @alekaizer 

 What would `registerForRemoteNotification` do?

**Comment by alekaizer** [comment]
@cruzach I considered using `expo-notification` but as I mentioned earlier, it seems to not support `topics` and I don't want to be fetching thousands of tokens and sending them messages, even it is in batch.

`registerForRemoteNotification` can be added to the expo-notification sdk to convert APNs tokens to FCM tokens([source](https://www.thepolyglotdeveloper.com/2017/06/apns-tokens-fcm-tokens-simple-http/))

I successfully converted 2 APNs tokens to FCM tokens, just received `InvalidAPNsCredentials` and investigating that at the moment

**Comment by cruzach** [comment]
> registerForRemoteNotification can be added to the expo-notification sdk to convert APNs tokens to FCM tokens(source)

I don't see a reason to add this as a method on top of `expo-notifications`, since you can do it yourself in JavaScript since Expo exposes the device push token. If you're receiving an invalid credentials error from FCM, that has to do with your FCM project setup, not with Expo, so I'm going to close this issue

**Comment by alekaizer** [comment]
@cruzach I don't think it can be on `expo-notifications` but on `expo-notifications-sdk` the server side package, so `expo-notifications` get you a APNs token and on the server you convert it to `fcm` token using `expo-notifications-sdk`

**Comment by cruzach** [comment]
If you're using FCM _directly_ you cannot use `expo-notifications-sdk`, because that is meant for the Expo Notifications service

We state in the docs:
> If you'd rather build a server that communicates with APNs and FCM directly, check out [this guide](https://docs.expo.dev/push-notifications/sending-notifications-custom/) (this is more complicated than using Expo's push notification service).

**Comment by alekaizer** [comment]
> If you're using FCM _directly_ you cannot use `expo-notifications-sdk`, because that is meant for the Expo Notifications service
> 
> We state in the docs:
> 
> > If you'd rather build a server that communicates with APNs and FCM directly, check out [this guide](https://docs.expo.dev/push-notifications/sending-notifications-custom/) (this is more complicated than using Expo's push notification service).

Gotcha! Thanks for the assistance though

**Comment by alekaizer** [comment]
@cruzach Sorry, one last thing, you didn't expand on the topic feature for `expo-notifications` is it possible and I'm missing it in the documentation, or it is just that I'll have to loop through each expo push token to send notification to each individual device?

**Comment by cruzach** [comment]
`expo-notifications` doesn't contain any code for topics right now

**Comment by alekaizer** [comment]
@cruzach any plans for it in the future?
Sending individual notifications isn't optimal as you know, or is there a workaround that you know of? (I'm considering switching to `expo-notification` and `expo-notification-sdk`, and as we discussed, topics is the only thing keeping me away from it)

**Comment by cruzach** [comment]
No plans right now, but if you open a feature request and it gains traction we might be able to add it to our roadmap: https://expo.canny.io/feature-requests?search=topic

**Comment by alekaizer** [comment]
@cruzach My guess is you've missed this, or you just got tired of me?

**Comment by cruzach** [comment]
I don't think we need to add specifics about the `@react-native-firebase` package to Expo's `expo-notifications` docs, that would be a little confusing. If you're using `@react-native-firebase`, you should refer to those docs.

There are definitely ways to register an iOS device with FCM, but it doesn't require any native code, just the device token, which is already available via `getDevicePushTokenAsync`. Other than that, all you need to do is refer to FCM's documentation 😄

**Comment by alekaizer** [comment]
@cruzach I see how it can be confusing, but we already have that with Android FCM configuration, right?
I can see on Expo's documentation how to set up Android with FCM and it was quite straightforward and worked like a charm, but for iOS nothing, it is like being trapped into using expo-server-sdk

What I'm suggesting is to add the configuration for iOS to work with expo-notification just like Android

- Adding the google-service Plist
- Add XYZ
- Configure XYZ

What I'm saying is, how can we achieve this
>There are definitely ways to register an iOS device with FCM, but it doesn't require any native code, just the device token, which is already available via getDevicePushTokenAsync. 

You can point us to the FCM documentation from Expo's documentation, so we don't have to pull our hairs 😆 
I've been trying really hard, even tried to do the project in RN, but as you know it's quite hectic while Expo was quite straight to the point, the only shenanigans with Expo now is the iOS Notification, the doc is really really complicated and it seems like my only alternative is to use expo-server-sdk and that one has that topic limitation because looping through devices tokens sounds so wrong to me

I don't know man, it just seems like the doc is missing something, we can make it less confusing, I'm begging you point me to the right direction cause I really tried to make iOS work to no avail 😢

**Comment by alekaizer** [comment]
@cruzach lemme check Expo documentation again maybe I missed something or my memory is outdated.
But my biggest question at this point is why did you guys elect to not add topics to `expo-notification` and `expo-server-sdk`?
why go the loop route?
